### PR TITLE
Consolidate requirements and repository configuration

### PR DESCRIPTION
**Pull request by rocco8773**
The primary goal here is to consolidate how we define package requirements (build, install, tests, docs, etc.).  Right now package requirements/dependencies are declared manually in many of the repository configuration files.  **After the refactor, dependencies should only be defined in the `requirements/*.txt` files and the `setup.cfg` file, with appropriate comments to direct developers.**  Any configuration file that needs dependencies should refer to the locations just mentioned.

This will most likely include housekeeping of the repository configuration files.

### CODE REVIEW

File: requirements/docs.txt [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# these are dependencies required to build package documentaiton
+# ought to mirror 'docs' under options.extras_require in config.cfg
+-r extras.txt
+numpydoc
+pillow
+sphinx <= 2.4.4
```
Comment (namurphy) [inline requirements/docs.txt:6]:
```suggestion
sphinx < 3.0
```
I'm not sure if we would want to do this instead on the chance that there's another bugfix release to 2.4.

File: setup.cfg [hunk 2]
```diff
@@ -8,7 +8,7 @@ long_description = file: README.md
 long_description_content_type = text/markdown
 license = BSD+Patent
 license_file = LICENSE.md
-url = http://www.plasmapy.org
+url = https://www.plasmapy.org
```
Comment (namurphy) [inline setup.cfg:11]: Good catch!

File: setup.cfg [hunk 3]
```diff
@@ -28,33 +29,50 @@ classifiers=
...
 include_package_data = True
-setup_requires = setuptools_scm
+setup_requires =
+  # PEP-518 encourages the use of pyproject.toml for defining
+  # build dependencies...they should be defined under build-system.requires
 install_requires =
+  # ought to mirror requirements/install.txt
   astropy >= 3.1
   colorama >= 0.3
+  matplotlib
```
Comment (namurphy) [inline setup.cfg:39]:
```suggestion
  matplotlib >= 2.0
```
We ought to specify a minimum version here too, though I'm pretty much guessing with 2.0.

File: requirements/install.txt [hunk 4]
```diff
@@ -0,0 +1,8 @@
+# these are dependencies required to use the package
+# ought to mirror 'instal_requires' under 'options' in config.cfg
+astropy >= 3.1
+colorama >= 0.3
```
Comment (namurphy) [inline requirements/install.txt:4]: Just FYI - #728 will remove colorama as a dependency

File: requirements/docs.txt (hunk 1 above)
Comment (rocco8773) [reply requirements/docs.txt:6]: I think they're officially on the `3.0.x` path. Just today they released `3.0.2` and hopefully the fix to `sphinx-automodapi` will be merged soon.

File: requirements/install.txt (hunk 4 above)
Comment (rocco8773) [reply requirements/install.txt:4]: That means I should do a little code review....😁

File: setup.cfg (hunk 3 above)
Comment (rocco8773) [reply setup.cfg:39]: I think it should be `matplotlib >= 3.2`.  That's when they started supporting python 3.8 (3.6+).  `2.0` still supports python 2.

File: requirements.txt [hunk 5]
```diff
@@ -0,0 +1,8 @@
+# all dependencies required to build, run, test, and doucment plasmapy
```
Comment (StanczakDominik) [inline requirements.txt:1]:
```suggestion
# all dependencies required to build, run, test, and document PlasmaPy
```

File: requirements/install.txt (hunk 4 above)
Comment (StanczakDominik) [reply requirements/install.txt:4]: Meanwhile #675 adds numba and xarray ;)

File: setup.cfg [hunk 6]
```diff
@@ -28,33 +29,52 @@ classifiers=
...
+  setuptools_scm
+  # for plasmapy.classes
   h5py
-  matplotlib
-diagnostics =
-  matplotlib
-tests =
-  pytest >= 5.1
-theory =
+  # for plasmapy.theory
```
Comment (StanczakDominik) [inline setup.cfg:54]:
```suggestion
  # for plasmapy.formulary 
```

File: setup.cfg [hunk 7]
```diff
@@ -28,33 +29,52 @@ classifiers=
...
   mpmath
   lmfit
+  # for plasmapy.utils.pytest_helpers
+  pytest >= 5.1
```
Comment (StanczakDominik) [inline setup.cfg:58]: I thought we fixed that... pytest should just be a dependency within `[tests]`, I think.

File: setup.cfg [hunk 8]
```diff
@@ -28,33 +29,52 @@ classifiers=
...
+  sphinx <= 2.4.4
   sphinx-automodapi
   sphinx-gallery
   sphinx_rtd_theme
   towncrier
+developer =
+  # install everything for developers
+  %(docs)s
+  %(extras)s
+  %(tests)s
```
[the remark this answers is not in the document]
Comment (namurphy) [reply setup.cfg:77]: Yeah, I noticed this too and had no idea we could do this.   I like it!

File: setup.cfg (hunk 3 above)
Comment (namurphy) [reply setup.cfg]: `matplotlib >= 3.2` sounds good to me!

**Comment by namurphy** [comment]
Astropy now requires Numpy version >= 1.16.0, so we probably ought to update that requirement in this PR to be consistent with them.

File: requirements/install.txt [hunk 9]
```diff
@@ -0,0 +1,8 @@
...
+matplotlib >= 3.2
+numpy >= 1.14
```
Comment (namurphy) [inline requirements/install.txt:6]:
```suggestion
numpy >= 1.16
```

**Review by namurphy — DISMISSED** [review]

File: requirements/install.txt (hunk 9 above)
Comment (StanczakDominik) [reply requirements/install.txt:6]: Does it, now? That's likely a sign of units issues going away!

File: setup.cfg (hunk 7 above)
Comment (namurphy) [reply setup.cfg:58]: There are a few things in `plasmapy.utils` that require pytest at the moment, but these will be mostly or entirely removed by #728.  After then, pytest should only be needed within `plasmapy.tests` and not `plasmapy.utils`.

**Comment by namurphy** [comment]
I just realized that requirements information will be needed for the CodeMeta file, so it's likely there will be one more place that will need to be updated...  This could be done just prior to every release, though.

**Comment by StanczakDominik** [comment]
I've been thinking that this would be worthwhile to prioritise. Would you mind if I picked this up from here?

**Comment by StanczakDominik** [comment]
@namurphy @rocco8773 a question. f1507583c92517f9750b43ccb1753dd4e2b5b730 claims that the `.zenodo.json` file is outdated and removes it. Can I ask for some context on that?

I would **gladly** remove it from the release guidelines :stuck_out_tongue_winking_eye:

**Comment by StanczakDominik** [comment]
One funny thing I realized with with the `%(extras)s` aprroach to extras in setup.cfg: it concatenates the requirements. So we get h5py a bunch of times. But then pip turns them into a set along the way, I think, so it should be fine.

**Comment by StanczakDominik** [comment]
@rocco8773 I'm not quite sure what you were referring to with the BaPSFlib setup I cleaned it up as much as I could for now, though.

**Comment by StanczakDominik** [comment]
Two issues that I'd rather leave for the next PR to touch these:

https://github.com/rocco8773/PlasmaPy/blob/7769ee263e7c70353fac9fa3496b1ff2befba90f/setup.cfg#L32-L34

This would probably simplify our conda packaging process a little.

https://github.com/rocco8773/PlasmaPy/blob/7769ee263e7c70353fac9fa3496b1ff2befba90f/tox.ini#L53-L90

I have not yet been able to simplify these, and would definitely like to. I suspect that there might be some way to specify "take the minimum of the range for these extras", maybe a pip install flag... but I don't know yet.

**Comment by rocco8773** [comment]
> @rocco8773 I'm not quite sure what you were referring to with the BaPSFlib setup. I cleaned it up as much as I could for now, though.

Look how I structured and commented  the requirements between the `./requirements/` directory, `setup.cfg`, and `pyproject.toml` in https://github.com/BaPSF/bapsflib/pull/47.  Basically I'm wanted link back comments so if we make a change to `setup.cfg` we know to also change `X`, and visa-versa.

**Comment by StanczakDominik** [comment]
> Look how I structured and commented the requirements between the `./requirements/` directory, `setup.cfg`, and `pyproject.toml` in [BaPSF/bapsflib#47](https://github.com/BaPSF/bapsflib/pull/47). Basically I'm wanted link back comments so if we make a change to `setup.cfg` we know to also change `X`, and visa-versa.

The appropriate comments were already there, just with `config.cfg` rather than `setup.cfg`! Anyway, looks like that's done.

**Review by StanczakDominik — APPROVED** [review]
This has stalled out long enough, and doesn't break stuff; in the interest of getting it through before we have to fix more merge conflicts, let's get it in!

**Pull request merged**